README.md
# Vulpes

[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![pypi version](https://img.shields.io/pypi/v/vulpes.svg)](https://pypi.python.org/pypi/vulpes)
[![Documentation Status](https://readthedocs.org/projects/vulpes/badge/?version=latest)](https://vulpes.readthedocs.io/en/latest/?badge=latest)
[![Downloads](https://static.pepy.tech/badge/vulpes)](https://pepy.tech/project/vulpes)

<img src="https://github.com/AdrienC21/vulpes/blob/main/logo_large.png?raw=true"  width=60% height=60%>

**Vulpes: Test many classification, regression models and clustering algorithms to see which one is most suitable for your dataset.**

Vulpes 🦊 is a Python package that allows you to test many models, whether you want to do classification, regression or clustering in your projects. It calculates many metrics for each model to compare them. It is highly customizable and it contains many features to save time building robust ML models.

If you like this project, please leave a star ⭐ on GitHub !

Alpha version.

Author & Maintainer: Adrien Carrel.

## Installation

Using pip:

```python
pip install vulpes
```

## Dependencies

vulpes requires:

- Python (>= 3.7)
- numpy (>= 1.22)
- pandas (>= 1.3.5)
- scikit-learn (>= 1.0.2)
- tqdm (>= 4.64.0)
- xgboost (>= 1.6.1)
- lightgbm (>= 3.3.2)

## Documentation

Link to the documentation: [https://vulpes.readthedocs.io/en/latest/](https://vulpes.readthedocs.io/en/latest/)

## Examples

General case, import one of the classes Classifiers, Regressions, Clustering from vulpes.automl, add some parameters to the object (optional), fit your dataset:

```python
from vulpes.automl import Classifiers
classifiers = Classifiers()
classifiers.fit(X, y)
```

More examples below and in notebooks in the folter **examples**.

### Classification

Fit many classification algorithms on the iris dataset from scikit-learn:

```python
import pandas as pd
from sklearn.datasets import load_iris
from vulpes.automl import Classifiers

dataset = load_iris()
X = pd.DataFrame(dataset["data"], columns=dataset["feature_names"])
y = dataset["target"]

classifiers = Classifiers(preprocessing="default")
df_models = classifiers.fit(X, y)
df_models
```

Analysis of each model using different metrics and repeated cross-validation by K-fold:

|                          Model | Balanced Accuracy | Accuracy | Precision |   Recall | F1 Score |    AUROC |    AUPRC | Micro avg Precision | Running time |
|-------------------------------:|------------------:|---------:|----------:|---------:|---------:|---------:|---------:|--------------------:|-------------:|
|   LinearDiscriminantAnalysis   |          0.977625 | 0.977333 |  0.978024 | 0.977625 | 0.976933 | 0.998161 | 0.996891 |            0.996940 |     4.372556 |
|  QuadraticDiscriminantAnalysis |          0.973219 | 0.973333 |  0.975460 | 0.973219 | 0.973162 | 0.999063 | 0.997595 |            0.997634 |     4.470590 |
|      LogisticRegressionCV      |          0.961609 | 0.961333 |  0.964101 | 0.961609 | 0.960668 | 0.997218 | 0.993264 |            0.993375 |    12.895212 |
|               SVC              |          0.961287 | 0.960000 |  0.962045 | 0.961287 | 0.959960 | 0.996825 | 0.994421 |            0.994510 |     4.437862 |
|     RandomForestClassifier     |          0.957220 | 0.956000 |  0.959982 | 0.957220 | 0.955394 | 0.993473 | 0.990367 |            0.989958 |    10.645725 |
|           GaussianNB           |          0.957169 | 0.954667 |  0.956188 | 0.957169 | 0.954521 | 0.993825 | 0.990463 |            0.990619 |     4.345500 |
|      ExtraTreesClassifier      |          0.956438 | 0.956000 |  0.958665 | 0.956438 | 0.955157 | 0.995156 | 0.991795 |            0.991704 |    10.440453 |
|       LogisticRegression       |          0.956094 | 0.954667 |  0.957273 | 0.956094 | 0.954427 | 0.997726 | 0.994765 |            0.994848 |     5.691309 |
|   GradientBoostingClassifier   |          0.955871 | 0.953333 |  0.956984 | 0.955871 | 0.953364 | 0.983221 | 0.967145 |            0.971317 |     9.005045 |
|          XGBClassifier         |          0.952846 | 0.950667 |  0.952745 | 0.952846 | 0.950324 | 0.985892 | 0.969083 |            0.972853 |     4.802282 |
|        BaggingClassifier       |          0.952712 | 0.950667 |  0.955214 | 0.952712 | 0.950581 | 0.985295 | 0.982312 |            0.971742 |     8.354026 |
|      KNeighborsClassifier      |          0.952699 | 0.950667 |  0.951586 | 0.952699 | 0.950683 | 0.990842 | 0.986716 |            0.980262 |     6.960091 |
|       AdaBoostClassifier       |          0.950432 | 0.946667 |  0.949250 | 0.950432 | 0.947114 | 0.988202 | 0.981889 |            0.977999 |     8.127254 |
|         LGBMClassifier         |          0.950009 | 0.948000 |  0.950426 | 0.950009 | 0.947522 | 0.991721 | 0.985483 |            0.985704 |     5.063474 |
|         LabelSpreading         |          0.948757 | 0.945333 |  0.947960 | 0.948757 | 0.946091 | 0.988827 | 0.981177 |            0.981552 |     4.332253 |
| HistGradientBoostingClassifier |          0.948195 | 0.945333 |  0.949260 | 0.948195 | 0.945352 | 0.988212 | 0.976375 |            0.976866 |     7.706454 |
|        LabelPropagation        |          0.946091 | 0.944000 |  0.946373 | 0.946091 | 0.944250 | 0.990341 | 0.984098 |            0.984373 |     4.406253 |
|          MLPClassifier         |          0.944773 | 0.941333 |  0.945336 | 0.944773 | 0.942314 | 0.992075 | 0.985516 |            0.985762 |     7.662322 |
|     DecisionTreeClassifier     |          0.942681 | 0.941333 |  0.944493 | 0.942681 | 0.940183 | 0.957011 | 0.951111 |            0.908000 |     4.367503 |
|            LinearSVC           |          0.936713 | 0.936000 |  0.937548 | 0.936713 | 0.933929 | 0.989648 | 0.983251 |            0.983539 |     4.474272 |
|       ExtraTreeClassifier      |          0.933964 | 0.932000 |  0.934967 | 0.933964 | 0.931137 | 0.950473 | 0.943333 |            0.893289 |     4.336813 |
|          SGDClassifier         |          0.922581 | 0.918667 |  0.927593 | 0.922581 | 0.919651 | 0.981940 | 0.962839 |            0.963484 |     5.666082 |
|     CalibratedClassifierCV     |          0.894860 | 0.888000 |  0.896616 | 0.894860 | 0.887397 | 0.972231 | 0.957643 |            0.958332 |     5.699280 |
|           Perceptron           |          0.873581 | 0.865333 |  0.887799 | 0.873581 | 0.864172 | 0.976069 | 0.945789 |            0.946695 |     4.482433 |
|         NearestCentroid        |          0.854566 | 0.854667 |  0.854707 | 0.854566 | 0.849341 | 0.973214 | 0.963677 |            0.964257 |     5.783815 |
|         RidgeClassifier        |          0.843743 | 0.834667 |  0.848879 | 0.843743 | 0.831310 | 0.945148 | 0.920905 |            0.922219 |     4.415888 |
|        RidgeClassifierCV       |          0.841049 | 0.832000 |  0.846498 | 0.841049 | 0.828592 | 0.944421 | 0.919460 |            0.920816 |     4.484041 |
|           BernoulliNB          |          0.757425 | 0.758667 |  0.771867 | 0.757425 | 0.728847 | 0.883542 | 0.839397 |            0.823834 |     4.479535 |
|         DummyClassifier        |          0.333333 | 0.249333 |  0.083111 | 0.333333 | 0.132452 | 0.500000 | 0.379100 |            0.299444 |     4.396426 |
|                                |                   |          |           |          |          |          |          |                     |              |

Here, the "default" preprocessing pipeline has been used. It consists of SimpleImputer (median strategy) with a StandardScaler for the features and a OneHotEncoder for the categorical features.

### Regressions

Fit many regression algorithms:

```python
from sklearn.datasets import make_regression
from vulpes.automl import Regressions

X, y = make_regression(
          n_samples=100, n_features=4, random_state=42, noise=4.0,
          bias=100.0)

regressions = Regressions()
df_models = regressions.fit(X, y)
df_models
```

### Clustering

Fit many clustering algorithms on the iris dataset from scikit-learn:

```python
import pandas as pd
from sklearn.datasets import load_iris
from vulpes.automl import Clustering

dataset = load_iris()
X = pd.DataFrame(dataset["data"], columns=dataset["feature_names"])

clustering = Clustering()
df_models = clustering.fit(X)
df_models
```

### Fit a "best model"

We can automatically build a VotingClassifier or a VotingRegressor using the build_best_models method once the models are fitted.

```python
df_best = classifiers.build_best_models(X, y, nb_models=3)
df_best
```

|           Model | Balanced Accuracy | Accuracy | Precision |  Recall | F1 Score | Running time |
|----------------:|------------------:|---------:|----------:|--------:|---------:|-------------:|
| Voting (3-best) |           0.97508 | 0.974667 |  0.976034 | 0.97508 | 0.974447 |     11.82946 |

### Check missing data

```python
import pandas as pd
import numpy as np
df = pd.DataFrame([["a", "x"],
                   [np.nan, "y"],
                   ["a", np.nan],
                   ["b", np.nan]],
                  dtype="category",
                  columns=["feature1", "feature2"])
classifiers.missing_data(df)
```

| Total Missing | Percentage (%) | Accuracy |
|--------------:|---------------:|---------:|
|    feature2   |              2 |     50.0 |
|    feature1   |              1 |     25.0 |

## Testing

If you want to submit a pull request or if you want to test in local the package, you can run some tests with the library pytest by running the following command:

```python
pytest vulpes/tests/
```

## Why Vulpes?

Vulpes stands for: **V**ector (**U**n)supervised **L**earning **P**rogram **E**stimation **S**ystem.

Nah, I'm kidding, I just love foxes, they are cute! The most common and widespread species of fox is the red fox (Vulpes vulpes).

![alt text](https://github.com/AdrienC21/vulpes/blob/main/fox.jpg?raw=true)

## Acknowledgment

- Shankar Rao Pandala (and some contributors). Their package (Lazy Predict) has been an inspiration.

## License

[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.rst
Changelog
==================================

0.2.0 (2022/07/05)
--------------------

- More documentation, add the ability to modify hyperparameters for different pipeline elements.

0.1.0 (2022/07/01)
--------------------

- Version 1 of the Vulpes python package.


vulpes/automl/classifiers.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""classifiers.py: Class Classifiers
to test many classification models
"""

from .corevulpes import CoreVulpes
from ..utils.utils import (
    CUSTOM_SCORER_CLF,
    METRIC_NAMES,
    METRICS_TO_REVERSE,
    create_model_2,
)

import warnings
import numbers
from time import perf_counter
from typing import List, Dict, Any, Union, Tuple
from collections import defaultdict
from collections.abc import Iterable
from os.path import join as opj

import numpy as np
import pandas as pd
from tqdm import tqdm
from sklearn.pipeline import Pipeline
from sklearn.model_selection import (
    cross_validate,
    RepeatedKFold,
    _validation,
    train_test_split,
)

# define type Array_like
Array_like = Union[List, pd.DataFrame, pd.Series, np.ndarray, Any]


class Classifiers(CoreVulpes):
    """
    Object to train many classifiers.
    All the parameters are optionals and can be modified.

    Args:
        models_to_try (Union[str, List[Tuple[str, Any]]], optional):
            List of models to try. It can be either a string that corresponds
            to a predefined list of models ("all", ...) or it can be a list
            of tuple (name of a model, class of a model)
            (e.g. ("RandomForestClassifier",
                   sklearn.ensemble.RandomForestClassifier)).
            Defaults to "all" (train all the available classification
            algorithms).
        custom_scorer (Dict[str, Any], optional): metrics to calculate after
            fitting a model. Dictionary with pairs name:scorer where the
            scorer is created using the function make_scorer from sklearn.
            Defaults to CUSTOM_SCORER_CLF.
        preprocessing (Union[Pipeline, str], optional): preprocessing
            pipeline to use. It can be None (no preprocessing), a
            predefined preprocessing pipeline ("default", ...) or
            a Pipeline object from sklearn. Defaults to "default":
            it applies a OneHotEncoder to "category" and object features,
            and it applies a SimpleImputer (median strategy) and a
            StandardScaler to numerical features.
        use_cross_validation (bool, optional): whether or not we apply
            a cross-validation. Defaults to True.
        cv (Any, optional): cross-validation object. It can be a predefined
            cross-validation setting ("default", "timeseries", ...), an
            iterable object, a cross-validation object from sklearn, etc.
            Defaults to "default": it applies a StratifiedShuffleSplit
            if a groups object is given when applying the fit method,
            otherwise, it uses a RepeatedKFold. In both cases, n_splits
            is set to 5.
        test_size (float, optional): test of the size set when splitting.
            Defaults to 0.2.
        shuffle (bool, optional): whether or not the algorithm shuffle the
            sample when splitting the given dataset. Defaults to False.
        sort_result_by (str, optional): on which metric do you want to
            sort the final dataframe. Defaults to "Balanced Accuracy".
        ascending (bool, optional): sort the final dataframe in
            ascending order?. Defaults to False.
        save_results (bool, optional): if True, save the results in a csv
            file. Defaults to False.
        path_results (str, optional): path to use when saving the results.
            Defaults to "".
        additional_model_params (Dict[str, Any], optional): dictionary
            that contains parameters to be applied to each element of the
            pipeline. E.g. {"n_estimators": 100}, apply to all the
            preprocessing tasks and/or models that have the parameter
            n_estimators with the parameter n_estimators. Defaults to {}.
        random_state (int, optional): random state variable. Is applied
            to every model and elements of the pipeline. Defaults to 42.
        verbose (int, optional): if greater than 1, print the warnings.
            Defaults to 0.
    Examples:

        >>> import pandas as pd
        >>> from sklearn.datasets import load_iris
        >>> from vulpes.automl import Classifiers

        >>> dataset = load_iris()
        >>> X = pd.DataFrame(dataset["data"], columns=dataset["feature_names"])
        >>> y = dataset["target"]

        >>> classifiers = Classifiers()
        >>> df_models = classifiers.fit(X, y)
        >>> df_models
        | Model                       | Balanced Accuracy | Accuracy | ...
        |-----------------------------|-------------------|----------|------
        Model									
        LinearDiscriminantAnalysis	    0.977625	        0.977333   ...
        MLPClassifier	                0.973753	        0.973333   ...
        QuadraticDiscriminantAnalysis	0.973219	        0.973333   ...
        KNeighborsClassifier	        0.971702	        0.969333   ...
        ...                             ...                 ...        ...
    """
    def __init__(
        self,
        *,
        models_to_try: Union[str, List[Tuple[str, Any]]] = "all",
        custom_scorer: Dict[str, Any] = CUSTOM_SCORER_CLF,
        preprocessing: Union[Pipeline, str] = "default",
        use_cross_validation: bool = True,
        cv: Any = "default",
        test_size: float = 0.2,
        shuffle: bool = False,
        sort_result_by: str = "Balanced Accuracy",
        ascending: bool = False,
        save_results: bool = False,
        path_results: str = "",
        additional_model_params: Dict[str, Any] = {},
        random_state: int = 42,
        verbose: int = 0,
    ):
        super().__init__()
        self.task = "classification"
        self.models_to_try = self.predefined_list_models(models_to_try)
        self.custom_scorer = custom_scorer
        self.preprocessing = preprocessing
        self.use_cross_validation = use_cross_validation
        self.cv = cv
        self.test_size = test_size
        self.shuffle = shuffle
        self.sort_result_by = sort_result_by
        self.ascending = ascending
        self.save_results = save_results
        self.path_results = path_results
        self.additional_model_params = additional_model_params
        self.random_state = random_state
        self.verbose = verbose

        if not (self.verbose):
            warnings.filterwarnings("ignore")

    def fit(
        self,
        X: Array_like,
        y: Array_like,
        *,
        sample_weight: Array_like = None,
        groups: Array_like = None,
    ) -> pd.DataFrame:
        """
        Fit many models

        Args:
            X (Array_like): Input dataset
            y (Array_like): Target variable
            sample_weight (Array_like, optional): sample weights
                to apply when fitting. Defaults to None.
            groups (Array_like, optional): groups to use during
                cross validation to stratify. Defaults to None.

        Raises:
            ValueError: cross validation wrong type, or failed
            RuntimeError: Error when fitting a model

        Returns:
            pd.DataFrame: dataframe with the goodness-of-fit metrics
                evaluated for each model.

        Examples:

        """
        # Convert X to dataframe
        # (some preprocessing task, model, etc require this format)
        if not (isinstance(X, pd.DataFrame)):
            X = pd.DataFrame(X)

        # dictionary to store calculated values, model info, etc for each model
        metrics_dic = defaultdict(list)

        # Fit and calculate metrics for all the models
        for name, model in tqdm(self.models_to_try):
            print("\n" + name)
            top = perf_counter()

            # extend the class with a custom predict proba function
            # based on a decision function or pairwise distances
            if not (hasattr(model, "predict_proba")):
                model = create_model_2(model)
            else:
                model = model()
            model_name = name.lower()

            # Create the pipeline
            pipe = self.create_pipeline(model_name, model)

            # When available:
            # - Activate "probability" parameter to calculate more metrics
            # - maximize n_jobs
            # - fix a random_state
            # - change the loss for SGDClassifier to allow the
            # predict_proba method
            model_params = {}
            for pipe_name, pipe_elt in pipe.steps:
                pipe_elt_available_params = pipe_elt.get_params().keys()
                if "random_state" in pipe_elt_available_params:
                    rd_state = f"{pipe_name}__random_state"
                    model_params[rd_state] = self.random_state
                if "normalize" in pipe_elt_available_params:
                    model_params[f"{pipe_name}__normalize"] = False
                if "n_jobs" in pipe_elt_available_params:
                    model_params[f"{pipe_name}__n_jobs"] = -1
                if "probability" in pipe_elt_available_params:
                    model_params[f"{pipe_name}__probability"] = True
                for (
                    add_param_name,
                    add_param_val,
                ) in self.additional_model_params.items():
                    if add_param_name in pipe_elt_available_params:
                        model_params[
                            f"{pipe_name}__{add_param_name}"
                        ] = add_param_val
            # change the loss to allow multiclass and predict proba
            if name == "SGDClassifier":
                model_params[f"{model_name}__loss"] = "log"
            if model_params != {}:
                pipe.set_params(**model_params)

            # if sample_weight is a parameter of the fit function, update
            # fit parameters
            fit_params = {}
            for pipe_name, pipe_elt in pipe.steps:
                if hasattr(pipe_elt, "fit") and (
                    "sample_weight" in pipe_elt.fit.__code__.co_varnames
                ):
                    fit_params[f"{pipe_name}__sample_weight"] = sample_weight

            if self.use_cross_validation:
                if not (
                    hasattr(self.cv, "split")
                    or isinstance(self.cv, numbers.Integral)
                    or isinstance(self.cv, Iterable)
                    or isinstance(self.cv, str)
                ):
                    raise ValueError(
                        "Expected cv as an integer, cross-validation "
                        "object (from sklearn.model_selection), "
                        "iterable or valid string predefined cv"
                    )
                if isinstance(self.cv, str):
                    cv = self.predefined_cv(X, groups)
                else:
                    cv = self.cv
                try:
                    cv_model = cross_validate(
                        pipe,
                        X,
                        y,
                        cv=cv,
                        return_estimator=True,
                        n_jobs=-1,
                        fit_params=fit_params,
                        scoring=self.custom_scorer,
                    )
                except ValueError as e:
                    print(e)
                    print("Cross validation failed. If groups provided, ")
                    print(
                        "maybe can't stratify because some groups "
                        "represented in the dataset don't contains enough"
                        " samples."
                    )
                    print("Using RepeatedKFold instead.")
                    cv = RepeatedKFold(
                        n_splits=5, n_repeats=5, random_state=self.random_state
                    )
                    cv_model = cross_validate(
                        pipe,
                        X,
                        y,
                        cv=cv,
                        return_estimator=True,
                        n_jobs=-1,
                        fit_params=fit_params,
                        scoring=self.custom_scorer,
                    )
                except Exception as e:
                    raise RuntimeError(str(e))

                # store fitted models
                self.fitted_models_[name] = cv_model["estimator"]

                # add metrics to the lists

                # nan mean are used, because when folding,
                # a label can be missing
                # in a particular fold, thus creating nan values
                for metric_name in self.custom_scorer.keys():
                    print_metric_name = METRIC_NAMES.get(metric_name,
                                                         metric_name)
                    (
                        metrics_dic[print_metric_name].append(
                            np.nanmean(cv_model[f"test_{metric_name}"])
                        )
                    )
                metrics_dic["Model"].append(name)  # add name
                # add running time
                metrics_dic["Running time"].append(perf_counter() - top)
            else:
                try:
                    X_train, X_test, y_train, y_test = train_test_split(
                        X,
                        y,
                        test_size=self.test_size,
                        shuffle=self.shuffle,
                        stratify=groups,
                        random_state=self.random_state,
                    )
                    pipe.fit(X_train, y_train, **fit_params)
                    res = _validation._score(
                        pipe, X_test, y_test, self.custom_scorer,
                        error_score="raise"
                    )
                except Exception as e:
                    raise RuntimeError(f"Error when fitting: {e}")

                # store fitted models
                self.fitted_models_[name] = pipe

                # add metrics to the lists

                # nan mean are used, because when folding,
                # a label can be missing
                # in a particular fold, thus creating nan values
                for metric_name in self.custom_scorer.keys():
                    print_metric_name = METRIC_NAMES.get(metric_name,
                                                         metric_name)
                    (
                        metrics_dic[print_metric_name].append(
                            np.nanmean(res[metric_name])
                        )
                    )
                metrics_dic["Model"].append(name)  # add name
                # add running time
                metrics_dic["Running time"].append(perf_counter() - top)

        # reverse these metrics (because lower is better)
        # ex: rmse, mae, mape
        for metric_name in METRICS_TO_REVERSE:
            if metric_name in metrics_dic:
                reverse_metric = [-x for x in metrics_dic[metric_name]]
                metrics_dic[metric_name] = reverse_metric
        # create a dataframe with the results
        df_models = pd.DataFrame.from_dict(metrics_dic)
        df_models = df_models.sort_values(
            by=self.sort_result_by, ascending=self.ascending
        ).set_index("Model")
        self.df_models = df_models

        if self.save_results:  # save results
            df_models.to_csv(
                opj.path.join(self.path_results,
                              f"vulpes_results_{self.task}.csv")
            )
        return df_models
